lib/gs-plugin-loader.c

```diff
@@ -1915,6 +1915,11 @@ save_install_queue (GsPluginLoader *plugin_loader)
 	g_autoptr(GString) s = NULL;
 	g_autofree gchar *file = NULL;
 
+	/* Saving the install queue during unit tests could trip up other tests as
+	 * they are run in parallel */
+	if (g_getenv ("GS_UNIT_TESTS_SKIP_SAVING_INSTALL_QUEUE") != NULL)
+		return;
```

Comment:
I don’t think this is the right solution. Unit tests should be testing the code which is going to be run in production, not skipping various parts of it.

Would it work to allow the install queue path to be set in the test harness, either by passing it into libgnome-software as an argument, or by creating a per-test context directory using `G_TEST_OPTION_ISOLATE_DIRS`?

Reply:
Yeah testing part of the code is sub-optimal but it's better than testing none at all which is the status quo. I'll look further into `G_TEST_OPTION_ISOLATE_DIRS`; I think you're right that would solve it and maybe it's not too much work to include it here. I happen to have also just filed [an issue about it](https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1043).

Reply:
Actually it looks like `G_TEST_OPTION_ISOLATE_DIRS` isn't a perfect solution for this: it doesn't take effect until `g_test_run()` is called so in `gs_plugin_loader_setup()` the install queue would be loaded from the system location, and then we would write the queue to the per-test isolated directory while the tests are running. So it would only work out fine in the normal case where the system install queue is empty.

Reply:
So I guess instead we should add a way to override the install queue directory with a new `gs_plugin_loader_...` function (and separately someone should make use of `G_TEST_OPTION_ISOLATE_DIRS`).

Reply:
Actually using an env var would be more in line with what we're doing with e.g. `GS_SELF_TEST_CACHEDIR`

Reply:
Did that in the latest version of the branch

Reply:
> Actually using an env var would be more in line with what we're doing with e.g. `GS_SELF_TEST_CACHEDIR`

I don’t think `GS_SELF_TEST_CACHEDIR` is the right approach to copy. We need fewer env vars, not more (they’re not threadsafe, are global state where local state is more useful, and are hard to debug and log). The only reason that env var still exists is because I ran out of time to rearchitect the appstream plugin when I last poked the tests. (See the comment on, for example, line 1860 of `plugins/flatpak/gs-self-test.c`.)

See also commits 1536efc75bd4c554a2615a8e2eb56ce25e17ce60 and 9f1f5ba2c3e9fff25bac20512523f29a502a6c7c, which I just dug up.

I think it should be possible to copy the `load_install_queue()` call from `gs_plugin_loader_setup()` to `gs_plugin_loader_setup_again()`, which means it will be called at the start of each unit test, to re-load the install queue from that test’s tmp directory. It might need modifying to correctly free and drop the old install queue, but I guess that should be fairly straightforward.

Reply:
I dropped the patch that set the environment variable to control the install queue location, and cherry-picked most of your patches from [460](https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/460), which is sufficient to stop the tests from interfering with each other's install queues.